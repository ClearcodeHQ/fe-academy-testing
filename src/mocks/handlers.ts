import { rest } from 'msw';

interface UserResponse {
  id: number;
  fullname: string;
}

export const handlers = [
  rest.get<UserResponse>('https://run.mocky.io/v3/bf2b59c6-bd8f-4fbe-9fda-21e686930673', (_, res, ctx) => {
    return res(ctx.status(200), ctx.json({ id: 2, fullname: 'John Doe' }));
  }),
];
