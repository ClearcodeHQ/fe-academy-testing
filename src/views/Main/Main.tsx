import { Counter } from 'components/Counter';
import { Year } from 'components/Year';
import React from 'react';
import { useFetch } from 'utils/useFetch';

export const MainView = () => {
  const { loading, error, data } = useFetch('https://run.mocky.io/v3/bf2b59c6-bd8f-4fbe-9fda-21e686930673');

  const renderData = (): React.ReactNode => {
    if (loading) {
      return <span>Fetching data...</span>;
    }

    if (error) {
      return <span>{error}</span>;
    }

    return <code>{data}</code>;
  };

  return (
    <>
      <Year />
      <Counter />
      <p>
        {renderData()}
      </p>
    </>
  );
};
