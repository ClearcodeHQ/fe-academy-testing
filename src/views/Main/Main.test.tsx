import { render, screen } from '@testing-library/react';
import { server, rest } from 'mocks/server';
import React from 'react';
import { MainView } from './Main';

describe('Integration: MainView', () => {
  beforeAll(() => {
    jest.useFakeTimers();
    jest.setSystemTime(new Date('2022-01-01'));
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  it('should match the snapshot', () => {
    const { asFragment } = render(<MainView />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should fetch API data', async () => {
    render(<MainView />);

    const data = await screen.findByText('{"id":2,"fullname":"John Doe"}');

    expect(data).toBeInTheDocument();
  });

  it('should render error when API call fails', async () => {
    server.use(
      rest.get('https://run.mocky.io/v3/bf2b59c6-bd8f-4fbe-9fda-21e686930673', async (_, res, ctx) =>
        res.once(ctx.status(403), ctx.text('User does not exist' ))
      )
    );
    
    render(<MainView />);

    const error = await screen.findByText('User does not exist');

    expect(error).toBeInTheDocument();
  });
});
