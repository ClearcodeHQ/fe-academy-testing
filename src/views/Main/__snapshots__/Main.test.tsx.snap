// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration: MainView should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bczRLJ hHQYvn"
  >
    <span>
      Year: 2022
    </span>
  </div>
  <span>
    Counter
  </span>
  <button>
    increment
  </button>
  <button>
    decrement
  </button>
  <span>
    count: 0
  </span>
  <p>
    <span>
      Fetching data...
    </span>
  </p>
</DocumentFragment>
`;
